File: examples/gemini_vision_example.py
#!/usr/bin/env python3

import sys
import gi
gi.require_version('Gst', '1.0')
gi.require_version('GstVideo', '1.0') # Though not explicitly used after simplification, good to keep for Gst context
from gi.repository import Gst, GLib, GObject # GstVideo might not be needed directly
import signal
import os # For API key environment variable (optional)

class GeminiVisionApp:
    def __init__(self):
        # Initialize GStreamer
        Gst.init(None)

        # --- Configuration for Gemini Vision element ---
        # IMPORTANT: Replace with your actual API key or set GST_GEMINI_API_KEY environment variable
        self.api_key = os.environ.get("GST_GEMINI_API_KEY", "YOUR_API_KEY_HERE")
        self.prompt_text = "Describe this scene in one short sentence. Focus on the main subject."
        self.model_name = "gemini-2.0-flash" # Or another model name
        self.analysis_interval_seconds = 5.0
        self.output_as_metadata = False # Use signal for this example

        # New generationConfig parameters
        self.stop_sequences = ["Title", "Conclusion"] # Python list of strings
        self.temperature = 0.7
        self.max_tokens = 150
        self.top_p_val = 0.9
        self.top_k_val = 20
        # --- End Configuration ---

        # Create the pipeline
        self.pipeline = Gst.Pipeline.new("gemini-vision-pipeline")

        # Create elements
        self.source = Gst.ElementFactory.make("videotestsrc", "source")
        self.convert1 = Gst.ElementFactory.make("videoconvert", "converter1")
        # Assuming your plugin is now named 'geminivision'
        self.gemini = Gst.ElementFactory.make("geminivision", "gemini-vision-element")
        self.convert2 = Gst.ElementFactory.make("videoconvert", "converter2")
        self.sink = Gst.ElementFactory.make("autovideosink", "sink")

        # Check elements were created
        if not all([self.pipeline, self.source, self.convert1, self.gemini, self.convert2, self.sink]):
            print("Error: Not all elements could be created.")
            if not self.gemini:
                print("Failed to create 'geminivision' element. Ensure the plugin is installed and discoverable (GST_PLUGIN_PATH).")
            sys.exit(1)

        # Configure elements
        self.source.set_property("pattern", 0)  # 0 is the smtpe pattern

        if self.api_key == "YOUR_API_KEY_HERE" or not self.api_key:
            print("ERROR: API Key is not set or is a placeholder.")
            print("Please set the GST_GEMINI_API_KEY environment variable or edit the script.")
            sys.exit(1)

        self.gemini.set_property("api-key", self.api_key)
        self.gemini.set_property("prompt", self.prompt_text)
        self.gemini.set_property("model-name", self.model_name)
        self.gemini.set_property("analysis-interval", self.analysis_interval_seconds)
        self.gemini.set_property("output-metadata", self.output_as_metadata)

        # Set new generationConfig properties
        # For G_TYPE_STRV, GObject introspection usually expects a list/tuple of strings from Python
        self.gemini.set_property("stop-sequences", self.stop_sequences)
        self.gemini.set_property("temperature", self.temperature)
        self.gemini.set_property("max-output-tokens", self.max_tokens)
        self.gemini.set_property("top-p", self.top_p_val)
        self.gemini.set_property("top-k", self.top_k_val)

        # Connect to the description-received signal
        self.gemini.connect("description-received", self.on_description_received)

        # Add elements to pipeline
        self.pipeline.add(self.source)
        self.pipeline.add(self.convert1)
        self.pipeline.add(self.gemini)
        self.pipeline.add(self.convert2)
        self.pipeline.add(self.sink)

        # Link elements
        if not self.source.link(self.convert1):
            print("Error: Could not link source to convert1")
            sys.exit(1)
        if not self.convert1.link(self.gemini):
            print("Error: Could not link convert1 to gemini")
            sys.exit(1)
        if not self.gemini.link(self.convert2):
            print("Error: Could not link gemini to convert2")
            sys.exit(1)
        if not self.convert2.link(self.sink):
            print("Error: Could not link convert2 to sink")
            sys.exit(1)

        # Create a main loop
        self.loop = GLib.MainLoop()

        # Add watch for messages from the pipeline
        bus = self.pipeline.get_bus()
        bus.add_signal_watch()
        bus.connect("message", self.on_message)

        # Handle keyboard interrupt
        signal.signal(signal.SIGINT, self.handle_sigint)

    def on_description_received(self, element, description, buffer):
        pts_ns = buffer.pts # PTS in nanoseconds
        pts_str = f"{pts_ns // 1_000_000_000}.{pts_ns % 1_000_000_000:09d}" # Format as S.NS
        
        print("=================================")
        print(f"Frame time: {pts_str} (PTS: {pts_ns})") # Matches GStreamer time format more closely
        print(f"Description: {description}")
        print("=================================\n")
        sys.stdout.flush() # Ensure it prints immediately like fflush(stdout)
        # Python's GObject signal handlers should typically return None or not return explicitly
        # unless a specific return value is expected by the signal emission.
        # For "run-last" signals without a specific return type, not returning is fine.

    def on_message(self, bus, message):
        t = message.type
        if t == Gst.MessageType.ERROR:
            err, debug = message.parse_error()
            print(f"Error from {message.src.get_name()}: {err.message}")
            if debug:
                print(f"Debug info: {debug}")
            self.loop.quit()
        elif t == Gst.MessageType.EOS:
            print("End of stream")
            self.loop.quit()
        elif t == Gst.MessageType.STATE_CHANGED:
            if message.src == self.pipeline:
                old_state, new_state, pending_state = message.parse_state_changed()
                print(f"Pipeline state changed from {Gst.Element.state_get_name(old_state)} to {Gst.Element.state_get_name(new_state)}")

    def handle_sigint(self, sig, frame):
        print("Interrupt received, stopping...")
        self.stop()

    def run(self):
        # Start the pipeline
        print("Setting pipeline to PLAYING state...")
        ret = self.pipeline.set_state(Gst.State.PLAYING)
        if ret == Gst.StateChangeReturn.FAILURE:
            print("Unable to set the pipeline to the playing state.")
            print("Check for errors from the 'geminivision' element, especially API key.")
            # Attempt to get more detailed error from bus if pipeline failed to start
            bus = self.pipeline.get_bus()
            msg = bus.timed_pop_filtered(Gst.CLOCK_TIME_NONE, Gst.MessageType.ERROR | Gst.MessageType.EOS)
            if msg:
                self.on_message(bus, msg)
            sys.exit(1)

        print("Pipeline running...")
        print("Press Ctrl+C to quit")

        try:
            self.loop.run()
        except Exception as e:
            print(f"Error in main loop: {e}")
        finally:
            self.stop()

    def stop(self):
        if hasattr(self, 'pipeline') and self.pipeline:
            print("Cleaning up...")
            self.pipeline.set_state(Gst.State.NULL)
        if hasattr(self, 'loop') and self.loop.is_running():
            self.loop.quit()
        print("Pipeline stopped.")

if __name__ == "__main__":
    app = GeminiVisionApp()
    app.run()

File: examples/gemini_vision_example.c
// gcc -o gemini_example gemini_example.c `pkg-config --cflags --libs gstreamer-1.0`

#include <gst/gst.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

typedef struct _AppData {
    GstElement *pipeline;
    GMainLoop *loop;
} AppData;

// Signal handler for "description-received" signal
static void 
on_description_received(GstElement *element, gchar *description, GstBuffer *buffer, gpointer user_data) {
    GstClockTime pts = GST_BUFFER_PTS(buffer);
    printf("=================================\n");
    printf("Frame time: %" GST_TIME_FORMAT "\n", GST_TIME_ARGS(pts));
    printf("Description: %s\n", description);
    printf("=================================\n\n");
    fflush(stdout);
}

// Bus callback to handle pipeline events
static gboolean
bus_callback(GstBus *bus, GstMessage *message, gpointer user_data) { // user_data is AppData*
    AppData *data = (AppData *)user_data;

    switch (GST_MESSAGE_TYPE(message)) {
        case GST_MESSAGE_ERROR: {
            GError *error = NULL;
            gchar *debug_info = NULL;
            
            gst_message_parse_error(message, &error, &debug_info);
            g_printerr("Error from %s: %s\n", GST_OBJECT_NAME(message->src), error->message);
            g_printerr("Debug info: %s\n", debug_info ? debug_info : "none");
            
            g_error_free(error);
            g_free(debug_info);
            g_main_loop_quit(data->loop);
            break;
        }
        case GST_MESSAGE_EOS:
            g_print("End of stream\n");
            g_main_loop_quit(data->loop);
            break;
        case GST_MESSAGE_STATE_CHANGED:
            if (GST_MESSAGE_SRC(message) == GST_OBJECT(data->pipeline)) {
                GstState old_state, new_state, pending_state;
                gst_message_parse_state_changed(message, &old_state, &new_state, &pending_state);
                g_print("Pipeline state changed from %s to %s\n",
                        gst_element_state_get_name(old_state),
                        gst_element_state_get_name(new_state));
            }
            break;
        default:
            break;
    }
    return TRUE;
}

int main(int argc, char *argv[]) {
    GstElement *pipeline, *source, *converter1, *gemini, *converter2, *sink;
    GstBus *bus;
    AppData data; // Declare AppData instance

    // --- Configuration for Gemini Processor ---
    const gchar *env_api_key = getenv("GST_GEMINI_API_KEY");
    const gchar *hardcoded_api_key = "YOUR_API_KEY_HERE"; // Fallback if env var is not set
    const gchar *api_key_to_use = NULL;

    if (env_api_key && env_api_key[0] != '\0') {
        api_key_to_use = env_api_key;
        g_print("Using API key from GST_GEMINI_API_KEY environment variable.\n");
    } else {
        api_key_to_use = hardcoded_api_key;
        if (strcmp(hardcoded_api_key, "YOUR_API_KEY_HERE") == 0 || hardcoded_api_key[0] == '\0') {
            g_print("Using fallback API key from code, but it's a placeholder or empty.\n");
        } else {
            g_print("Using API key hardcoded in the application.\n");
        }
    }
    
    const gchar *prompt_text = "Describe this scene in one short sentence. Focus on the main subject.";
    const gchar *model = "gemini-1.5-flash-latest"; // Or another model name
    gdouble analysis_interval_seconds = 5.0;
    gboolean output_as_metadata = FALSE; // Use signal for this example

    // New generationConfig parameters
    const gchar *stop_sequences[] = {"Title", "Conclusion", NULL}; // Example stop sequences, NULL terminated
    gdouble temperature = 0.7;
    gint max_tokens = 150;
    gdouble top_p_val = 0.9;
    gint top_k_val = 20;
    // --- End Configuration ---


    // Initialize GStreamer
    gst_init(&argc, &argv);

    // Create the pipeline elements
    pipeline = gst_pipeline_new("gemini-test-pipeline");
    source = gst_element_factory_make("videotestsrc", "source");
    converter1 = gst_element_factory_make("videoconvert", "converter1");
    gemini = gst_element_factory_make("geminivision", "gemini");
    converter2 = gst_element_factory_make("videoconvert", "converter2");
    sink = gst_element_factory_make("autovideosink", "sink");

    // Check that all elements were created successfully
    if (!pipeline || !source || !converter1 || !gemini || !converter2 || !sink) {
        g_printerr("Failed to create one or more elements:\n");
        if (!pipeline) g_printerr("  Pipeline is NULL\n");
        if (!source) g_printerr("  Source is NULL\n");
        if (!converter1) g_printerr("  Converter1 is NULL\n");
        if (!gemini) g_printerr("  Gemini (geminivision) is NULL. Check plugin path (GST_PLUGIN_PATH) and plugin installation.\n");
        if (!converter2) g_printerr("  Converter2 is NULL\n");
        if (!sink) g_printerr("  Sink is NULL\n");
        return 1;
    }

    // Initialize AppData
    data.pipeline = pipeline;
    data.loop = g_main_loop_new(NULL, FALSE);
    if (!data.loop) {
        g_printerr("Failed to create main loop\n");
        gst_object_unref(pipeline);
        return 1;
    }

    // Configure the video source
    g_object_set(G_OBJECT(source), "pattern", 18, NULL);  // 18 is GST_VIDEO_TEST_SRC_SMPTE

    // Configure the Gemini vision
    // IMPORTANT: Check if api_key_to_use is valid, otherwise the plugin will error out on start
    if (api_key_to_use == NULL || api_key_to_use[0] == '\0' || strcmp(api_key_to_use, "YOUR_API_KEY_HERE") == 0) {
        g_printerr("API Key is not set, is empty, or is a placeholder.\n");
        g_printerr("Please set the GST_GEMINI_API_KEY environment variable or edit the gemini_vision_example.c to provide a valid API key.\n");
        gst_object_unref(pipeline);
        g_main_loop_unref(data.loop);
        return 1;
    }

    g_object_set(G_OBJECT(gemini),
                 "api-key", api_key_to_use,
                 "prompt", prompt_text,
                 "model-name", model,
                 "analysis-interval", analysis_interval_seconds,
                 "output-metadata", output_as_metadata,
                 // Set new generationConfig properties
                 "stop-sequences", stop_sequences, // Pass the gchar** array directly
                 "temperature", temperature,
                 "max-output-tokens", max_tokens,
                 "top-p", top_p_val,
                 "top-k", top_k_val,
                 NULL);

    // Connect the "description-received" signal
    g_signal_connect(gemini, "description-received", G_CALLBACK(on_description_received), NULL);

    // Add all elements to the pipeline
    gst_bin_add_many(GST_BIN(pipeline), source, converter1, gemini, converter2, sink, NULL);

    // Link the elements
    if (!gst_element_link_many(source, converter1, gemini, converter2, sink, NULL)) {
        g_printerr("Elements could not be linked\n");
        gst_object_unref(pipeline);
        g_main_loop_unref(data.loop);
        return 1;
    }

    // Set up the bus watch
    bus = gst_pipeline_get_bus(GST_PIPELINE(pipeline));
    gst_bus_add_watch(bus, bus_callback, &data); // Pass address of AppData
    gst_object_unref(bus);

    // Start playing
    g_print("Setting pipeline to PLAYING state...\n");
    GstStateChangeReturn ret = gst_element_set_state(pipeline, GST_STATE_PLAYING);
    if (ret == GST_STATE_CHANGE_FAILURE) {
        g_printerr ("Unable to set the pipeline to the playing state.\n");
        gst_object_unref (pipeline);
        g_main_loop_unref(data.loop);
        return -1;
    }


    // Wait until an error occurs or EOS
    g_print("Running...\n");
    g_print("Press Ctrl+C to quit\n");
    g_main_loop_run(data.loop);

    // Clean up
    g_print("Cleaning up...\n");
    gst_element_set_state(pipeline, GST_STATE_NULL);
    gst_object_unref(GST_OBJECT(pipeline));
    g_main_loop_unref(data.loop);

    return 0;
}